Add upe integration for the Checkout Block

reykjalin opened the pull request:
# Changes proposed in this Pull Request:

Fixes #1667 
Fixes #1787 

- When the new checkout experience (UPE) is enabled in settings, use the UPE element in the Checkout Block.


Parts of this PR cannot be tested and figured out until the UPE implementation in general supports Subscriptions and Pre-Orders.

I still expect parts of this PR to break once some of the in-progress PRs are merged, but this is ready to review with the current status in `develop`.


# Testing instructions

Go through the following critical flows with the UPE experience enabled and using the Checkout block on the checkout page.

## With UPE disabled

We've run into some issues when UPE has been disabled, so please run through the following critical flows with UPE disabled, using both the Blocks and Shortcode versions of the checkout flow.

* [Checkout with normal credit card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-normal-credit-card)
* [Save card at checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#save-card-at-checkout)
* [Checkout with saved card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-saved-card)
* [Checkout with SCA card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-sca-card)
* [Checkout failures (with various cards)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-failures-with-various-cards)
* [Checkout with Payment Request](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-payment-request)
* [Blocks: Checkout with normal credit card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-normal-credit-card)
* [Blocks: Checkout with SCA card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-sca-card)
* [Blocks: Checkout failures (with various cards)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-failures-with-various-cards)
* [Blocks: Save card at checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-save-card-at-checkout)
* [Blocks: Checkout with saved card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-saved-card)
* [Blocks: Checkout with Payment Request](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-payment-request-button)

## Gutenberg

- [Blocks: Create a Checkout page with the Checkout Block](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-create-a-checkout-page-with-the-checkout-block)

## Adding/deleting payment methods

This PR makes some changes to the structure of the JS and shifts many functions to new files. Please make sure the non-checkout based functionality that uses the UPE element also works:

* [Add new payment method](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#add-new-payment-method#add-new-payment-method)
* [Delete payment method](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#add-new-payment-method#delete-payment-method)

## Card payments

Please make sure to test both the Shortcode and Blocks-based checkout flows.

### Shortcode flow

* [Checkout with normal credit card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-normal-credit-card)
* [Save card at checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#save-card-at-checkout)
* [Checkout with saved card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-saved-card)
* [Checkout with SCA card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-sca-card)
* [Checkout failures (with various cards)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-failures-with-various-cards)
* [Checkout with Payment Request](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#checkout-failures-with-various-cards#checkout-with-payment-request)

### Blocks flow

* [Blocks: Checkout with normal credit card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-normal-credit-card)
* [Blocks: Checkout with SCA card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-sca-card)
* [Blocks: Checkout failures (with various cards)](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-failures-with-various-cards)
* [Blocks: Save card at checkout](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-save-card-at-checkout)
* [Blocks: Checkout with saved card](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-saved-card)
* [Blocks: Checkout with Payment Request](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-payment-request-button)


## Payment Request Buttons

Depends on:

- [x] #1908 

* [Blocks: Checkout with Payment Request](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-checkout-with-payment-request-button)

## Subscriptions

Depends on:

- [ ] #1836 

* [Blocks: Purchase subscription product](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-purchase-subscription-product)
* [Blocks: Purchase free trial subscription](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-purchase-free-trial-subscription)
* [Blocks: Purchase multiple subscriptions](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#blocks-purchase-multiple-subscriptions)
* [Renew subscription manually](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#renew-subscription-manually)
* [Renew subscription automatically](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#renew-subscription-automatically)
* [Renew subscription](https://github.com/woocommerce/woocommerce-gateway-stripe/wiki/Testing-instructions-for-critical-flows#renew-subscription)

-------------------
- [x] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [x] Did you make changes, or create a **new .js file**? If **Gruntfile.js** exists in the repo, make sure to run `grunt`.

Commit:
Update npm packages

This updates the Stripe JS dependencies and adds a webpack resolver for
eslint

Commit:
Use the webpack resolver for eslint

This is so ESLint understands the wcstripe alias

Commit:
Fix issue where ajax endpoint was duplicated

When array_merge_recursive is used properties with the same name are
added to an array. This made it so that the ajax_url property was
duplicated in the frontend causing API requests to fail.

By using array_replace_recursive instead we fix this issue.

Commit:
Render UPE element in the Checkout Block

Payments don't work, but the element is there!

Commit:
Dequeue classic UPE JS to prevent JS conflicts

This was causing the blocks integration to fail for some reason and
dequeuing the classic JS fixes the issue.

Commit:
Add support for payment processing

This seems to be a complete integration but I haven't tested any edge
cases. We'll see how that goes later.

Both card failures and 3DS failures test well.

Commit:
Add todo for removing some temporary compat code

Commit:
Add missing fields to checkout response

Commit:
Add support for saved payment methods

reykjalin commented on client/blocks/upe/confirm-card-payment.js:
```diff
@@ -0,0 +1,47 @@
+	const { redirect, payment_method: paymentMethod } = paymentDetails;
+
+	try {
+		const confirmation = api.confirmIntent(
+			redirect,
+			shouldSavePayment ? paymentMethod : null
+		);
```
This works even though `paymentMethod` is always `undefined` here, probably since we only need the payment method ID if we're actually saving the payment method? So maybe we can just replace the `shouldSavePayment ? paymentMethod : null` with `null` or `undefined`?

I don't fully understand what the API expects here, so I'm at a bit of a loss here; this is all copied from WCPay.

I'm honestly not really sure what this parameter is for

Commit:
Add Blocks supports array when UPE is in use

reykjalin commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -356,6 +411,11 @@ public function get_supported_features() {
 			$gateway = $gateways['stripe'];
 			return array_filter( $gateway->supports, [ $gateway, 'supports' ] );
 		}
+
+		if ( WC_Stripe_Feature_Flags::is_upe_checkout_enabled() && empty( $gateways ) ) {
+			$gateway = new WC_Stripe_UPE_Payment_Gateway();
+			return array_filter( $gateway->supports, [ $gateway, 'supports' ] );
+		}
```
This seems extremely backwards, but without this the supports array is empty when using UPE-based checkout. That would be a problem when we add support for Subscriptions and Pre-Orders when using UPE.

Ideally this should be removed, and I'm hoping the UPE gateway will become part of the `$gateways` array once #1856 is merged.

diegocurbelo replied:
Can't reproduce this behavior, debugged this for a while, and the UPE Gateway is also called `stripe` so it's always set and returns in the first if.

We can however change it to not access the property `payment_gateways ` directly:
```
$gateways = WC()->payment_gateways()->get_available_payment_gateways();
```

reykjalin replied:
Yep, I can no longer reproduce this after #1855 was merged, so I've removed this code 🙂

reykjalin commented on .eslintrc.js:
```diff
@@ -19,5 +19,6 @@ module.exports = {
 			'@wordpress/element',
 			'@wordpress/data',
 		],
+		'import/resolver': 'webpack',
```
This is to fix an issue with ESLint. Should probably be part of #1854

reykjalin commented on package.json:
```diff
@@ -50,6 +50,7 @@
     "colors": "^1.4.0",
     "config": "^1.24.0",
     "cross-env": "^5.1.6",
+    "eslint-import-resolver-webpack": "0.13.1",
```
This is to fix an issue with ESLint. Should probably be part of #1854

reykjalin commented on client/blocks/upe/fields.js:
```diff
@@ -0,0 +1,214 @@
+		clientSecret,
+		business: { name: getStripeServerData()?.accountDescriptor },
+		fields: {
+			billingDetails: {
+				name: 'never',
+				email: 'never',
+				phone: 'never',
+				address: {
+					country: 'never',
+					line1: 'never',
+					line2: 'never',
+					city: 'never',
+					state: 'never',
+					postalCode: 'never',
+				},
+			},
```
I'm pretty sure these shouldn't all be set to `'never'`, but instead use the `enabledBillingFields` setting we embed from the PHP side. Not sure how that plays with the Blocks integration though; that's something we need to look into.

reykjalin commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -60,6 +60,63 @@ public function is_active() {
 	 * @return array
 	 */
 	public function get_payment_method_script_handles() {
+		if ( WC_Stripe_Feature_Flags::is_upe_checkout_enabled() ) {
+			$this->register_upe_payment_method_script_handles();
+		} else {
+			$this->register_classic_payment_method_script_handles();
+		}
```
I think this is a better approach to loading the relevant JS because it allows us to load smaller JS bundles.

The alternative would be to have the JS decide whether the UPE or "Classic" experience should be loaded, but that would mean a bigger JS bundle since the JS would contain both the "Classic" and UPE experiences; roughly half of the bundle would go unused in this case.

The downside of this approach is that the built plugin assets will be a bit bigger so merchants will need to download a slightly bigger plugin archive when installing/updating the plugin. I think the trade-off here is worth it though; a lighter (and thus faster) checkout experience is always worth it.

That said, I'm still open to suggestions on this 🙂

reykjalin commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -60,6 +60,63 @@ public function is_active() {
+			'woocommerce-gateway-stripe'
+		);
+
+		// TODO: remove once script conflict has been resolved.
+		wp_dequeue_script( 'wc-stripe-upe-classic' );
```
I'm hoping we can remove this once #1855 has been merged since the classic UPE JS (`upe_classic.js`) has some sort of conflict with the Blocks integration (`upe_blocks.js`).

**Note:** this (very simple) workaround breaks the shortcode checkout flow, so we should not merge this PR until this has been resolved.

reykjalin replied:
This has been removed since #1855 has been merged.

The issue itself also turned out to be a bit more nuanced, as can be seen with https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1825/commits/a9774f9f6745f8effaa9d3fb0df6d133248999fa

diegocurbelo commented on client/blocks/upe/fields.js:
```diff
@@ -0,0 +1,214 @@
+			} catch ( error ) {
+				setErrorMessage(
+					error?.message ??
+						'There was an error loading the payment gateway'
```
We should use `__()`:
```
__( 'There was an error loading the payment gateway', 'woocommerce-gateway-stripe' )
```

reykjalin replied:
Fixed in c8c04c83ca6989971a45ffea9d9079eee7fbc177

diegocurbelo commented on client/blocks/upe/fields.js:
```diff
@@ -0,0 +1,214 @@
+				if ( ! isUpeComplete ) {
+					return {
+						type: 'error',
+						message: 'Your payment information is incomplete.',
```
We should use `__()`:
```
message: __( 'Your payment information is incomplete.', 'woocommerce-gateway-stripe' ),
```

reykjalin replied:
Fixed in c8c04c83ca6989971a45ffea9d9079eee7fbc177

diegocurbelo commented on client/blocks/upe/fields.js:
```diff
@@ -0,0 +1,214 @@
+					return {
+						type: 'error',
+						message:
+							'This payment method can not be saved for future use.',
```
We should use __():
```
__( 'This payment method can not be saved for future use.', 'woocommerce-gateway-stripe' ), 
```

reykjalin replied:
Fixed in c8c04c83ca6989971a45ffea9d9079eee7fbc177

diegocurbelo commented on includes/class-wc-stripe-blocks-support.php:
```diff
@@ -60,6 +60,63 @@ public function is_active() {
+	/**
+	 * Registers the classic JS scripts.
+	 */
+	private function register_classic_payment_method_script_handles() {
```
I think that `classic` is a bit confusing here, generally, we refer to the shortcode checkout version as classic, like in `upe_classic.js`, and here we are registering the normal (non-UPE) version of the blocks integration.

But I don't have a better idea than `register_non_upe_payment_method_script_handles` or `register_legacy_payment_method_script_handles `😞 .

Maybe the problem es using `classic` for the shortcode and we need to change that 🤷🏼

reykjalin replied:
I like `legacy`; seems pretty accurate since the aim is to _eventually_ switch entirely to UPE 🤔 

Changed it to `legacy` in a67046de

diegocurbelo reviewed (changes requested):
Code looks good, and test very well 👏🏼 👏🏼 . Just added a couple of comments about missing translations.

I'm going to double check the `wp_dequeue_script( 'wc-stripe-upe-classic' );` that I can't reproduce.

Commit:
add translations for strings

Commit:
rename function: classic -> legacy

Commit:
Remove workaround for Block checkout JS conflict

It's been fixed in develop, so there's no need to keep the workaround.

Commit:
Remove unnecessary duplicate Stripe JS enqueue

This used to be a problem until #1856 was merged.

Commit:
Remove workaround for supports array

This used to be necessary until #1856 was merged.

Commit:
Add separate function for Blocks config

The getStripeServerData() function would return the Classic UPE config
even if the Checkout block was being used on the current page. This
caused an issue in the Checkout Block integration where the Stripe
payment method would not be made available.

The fix was to create a new getStripeServerData() function specific to
the Blocks integration and use that everywhere in client/blocks.

reykjalin commented on client/stripe-utils/utils.js:
```diff
@@ -23,15 +23,13 @@ import { errorTypes, errorCodes } from './constants';
-			? wc_stripe_upe_params // eslint-disable-line camelcase
-			: wc?.wcSettings?.getSetting( 'stripe_data', null );
-	if ( ! stripeServerData ) {
+	// Classic checkout.
+	// eslint-disable-next-line camelcase
+	if ( ! wc_stripe_upe_params ) {
 		throw new Error( 'Stripe initialization data is not available' );
 	}
-	return stripeServerData;
+	// eslint-disable-next-line camelcase
+	return wc_stripe_upe_params;
 };
```
This function is now exclusively used for the UPE checkout, since it was causing a conflict with the Blocks integration. See the commit message in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1825/commits/a9774f9f6745f8effaa9d3fb0df6d133248999fa for more details.

reykjalin commented on client/blocks/utils.js:
```diff
@@ -0,0 +1,14 @@
+ */
+import { getSetting } from '@woocommerce/settings';
+
+export const getStripeServerData = () => {
+	const stripeServerData = getSetting( 'stripe_data', null );
+
+	if ( ! stripeServerData ) {
+		throw new Error( 'Stripe initialization data is not available' );
+	}
+
+	return stripeServerData;
+};
```
This function is now exclusively used for the Blocks integration, since the `getStripeServerData()` function in `stripe-utils` was causing a conflict with the UPE integration. See the commit message in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1825/commits/a9774f9f6745f8effaa9d3fb0df6d133248999fa for more details.

I decided to go with the same name since it serves the same purpose as the `getStripeServerData()` function in `stripe-utils`, but that may be confusing. Would it make more sense to call this function `getBlocksConfiguration()` (or similar) so it doesn't get confused with the other `getStripeServerData()` function?

diegocurbelo replied:
I like the idea of having different names... but in that case should we also rename the classic `getStripeServerData()` (the one in `client/stripe-utils/utils.js`) `getClassicConfiguration()`?

reykjalin replied:
I think so? But I'm worried about messing up other people's PRs that way, so I think we should open an issue for that 🤔 

For now I can change the Blocks integration function to `getBlocksConfiguration()` to reduce the ambiguity 👍

reykjalin replied:
Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/1825/commits/a9e2c1955077f82046623692bdea54039c9c94a4

Commit:
Rename blocks config function

Renamed to getBlocksConfiguration() from getStripeServerData() to reduce
ambiguity with the UPE version of the same function that's also called
getStripeServerData(). Now the 2 functions are entirely distinct.

Commit:
Add the blocks container around the UPE element

This ensures proper spacing around the UPE element so the Save card
button doesn't overlap the inputs.

Commit:
Use right config for blocks API functions

Commit:
Respect enabled billing fields for Blocks UPE

Commit:
Use the config for PRB checkout

Commit:
Ensure stripe JS is loaded for UPE support

Commit:
Fix linting issues following rebase

reykjalin commented:
@diegocurbelo the PR is ready for another review, but unfortunately parts of it can't be tested because of dependencies on other PRs and issues.

I think it's probably worth it to review the code and make sure the tests that work pass so there's less to do next time, but it may also be better to simply wait. I'll ping you again no matter what you decide to do once the PR can be reviewed more thoroughly 🙂

Commit:
Isolate blocks configuration from UPE checkout

This is necessary to prevent breaking the shortcode UPE checkout while
the Blocks checkout works as intended and vice versa.

Commit:
Fix Blocks integration not loading when UPE disabled

Commit:
Use fetch instead of jquery.post

This will set us up to lower the cost of enqueuing the checkout JS since
we're no longer depending on jQuery.

reykjalin commented on client/api/blocks.js:
```diff
@@ -0,0 +1,77 @@
+import $ from 'jquery';
```
This file was added to separate the UPE/combined API calls entirely from the Blocks API calls.

reykjalin commented on client/blocks/utils.js:
```diff
@@ -0,0 +1,57 @@
+import { getSetting } from '@woocommerce/settings';
```
These are utility functions exclusive to the Blocks integration.

Commit:
Merge branch 'develop' into 1787-add-upe-integration-for-blocks

Commit:
Remove unnecessary eslint ignore

diegocurbelo commented on client/stripe-utils/utils.js:
```diff
@@ -16,60 +16,13 @@ import { errorTypes, errorCodes } from './constants';
-			? wc_stripe_upe_params // eslint-disable-line camelcase
-			: wc?.wcSettings?.getSetting( 'stripe_data', null );
-	if ( ! stripeServerData ) {
+	// Classic checkout.
+	// eslint-disable-next-line camelcase
+	if ( ! wc_stripe_upe_params ) {
 		throw new Error( 'Stripe initialization data is not available' );
```
Reverting this solves the `wc_stripe_upe_params is not defined` error
<img width="658" alt="Screen Shot 2021-09-22 at 17 12 52" src="https://user-images.githubusercontent.com/407542/134418373-08328fe1-5b64-43f6-aa66-573dc507b25e.png">
:

Commit:
Pass configuration into WCStripeAPI

This removes the dependency on the external configuration which will
ultimately make the class more testable as well as being easier across
different use cases going forward.

Commit:
Merge branch 'develop' into 1787-add-upe-integration-for-blocks

Commit:
Merge branch 'develop' into 1787-add-upe-integration-for-blocks

Commit:
Merge branch 'develop' into 1787-add-upe-integration-for-blocks
